### PR TITLE
openblas: test build with GCC 12

### PR DESCRIPTION
**Pull request by carlocab**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

The test for openblas failed in #106755. Let's try to reproduce it by
building against our existing `gcc@12`.

### CODE REVIEW

**Comment by fxcoudert** [comment]
```
./test: error while loading shared libraries: libopenblas.so.0: ELF load command address/offset not properly aligned
```

I'm not sure why `strip` would produce an invalid ELF file, though.

File: Formula/openblas.rb [hunk 1]
```diff
@@ -41,11 +42,35 @@ def install
...
 
     # Must call in two steps
-    system "make", "CC=#{ENV.cc}", "FC=gfortran", "libs", "netlib", "shared"
+    system "make", "CC=#{ENV.cc}", "FC=gfortran-12", "libs", "netlib", "shared"
     system "make", "PREFIX=#{prefix}", "install"
 
     lib.install_symlink shared_library("libopenblas") => shared_library("libblas")
     lib.install_symlink shared_library("libopenblas") => shared_library("liblapack")
+
+    (buildpath/"test.c").write <<~EOS
```
Comment (carlocab) [inline Formula/openblas.rb:51]:
```suggestion
    (buildpath/"brewtest.c").write <<~EOS
```

File: Formula/openblas.rb [hunk 2]
```diff
@@ -41,11 +42,35 @@ def install
...
+        for (i = 0; i < 9; i++)
+          printf("%lf ", C[i]);
+        printf("\\n");
+        if (fabs(C[0]-11) > 1.e-5) abort();
+        if (fabs(C[4]-21) > 1.e-5) abort();
+        return 0;
+      }
+    EOS
+    system ENV.cc, "test.c", "-I#{include}", "-L#{lib}", "-lopenblas", "-o", "test"
+    system "./test"
```
Comment (carlocab) [inline Formula/openblas.rb:73]:
```suggestion
    system ENV.cc, "brewtest.c", "-I#{include}", "-L#{lib}", "-lopenblas", "-o", "brewtest"
    system "./brewtest"
```

**Comment by carlocab** [comment]
@fxcoudert It is not a bottling issue, unfortunately (or fortunately?). The same test fails when run before bottling.

CC @Homebrew/core for ideas.

**Comment by fxcoudert** [comment]
Might be a gcc 11 -> 12 regression, or a binutils bug somewhere, if we can build in the same exact sources with gcc 11.

**Comment by carlocab** [comment]
I'm sceptical it is a GCC bug. Arch and Gentoo build with GCC 12 these days, and I don't see them carrying any patches for openblas.

**Comment by fxcoudert** [comment]
I can reproduce locally the issue locally with GCC 11, i.e. without this PR. Yet the test passed in https://github.com/Homebrew/homebrew-core/pull/95513 so… not sure what is happening.

**Comment by carlocab** [comment]
My guess is that it's from a change in `brew`.

**Comment by carlocab** [comment]
I can reproduce this, but only using a Homebrew GCC. This happens even without using `brew` to build from a formula.

I first used an Ubuntu-built toolchain on `ubuntu-18.04` (GCC 7), `ubuntu-20.04` (GCC 9), and `ubuntu-22.04` (GCC 11). This worked fine: https://github.com/carlocab/workflow-test/actions/runs/2774782818

Then, I used the same Ubuntu versions with various versions of Homebrew GCC. This leads to the same error with some versions of Homebrew GCC: https://github.com/carlocab/workflow-test/actions/runs/2774820275

It does work on Ubuntu 22.04 with GCC 10, though. Interestingly, the earlier versions of Ubuntu did not work with the same GCC.

**Comment by carlocab** [comment]
Or, actually, it looks like it'll work with GCC 7+ on Ubuntu 22.04?

**Comment by fxcoudert** [comment]
- Homebrew's GCC is configured in a rather straightforward way, except for the extensive rewrite of the specs. But I'm not seeing how that could be the cause of this particular error. ping @iMichka who wrote those
- Homebrew's GCC comes with a specific binutils version, that might be the cause. @carlocab can you test a system GCC with Homebrew's binutils in the path?

**Comment by carlocab** [comment]
> can you test a system GCC with Homebrew's binutils in the path?

https://github.com/carlocab/workflow-test/actions/runs/2775077969

**Comment by carlocab** [comment]
Yep, that did it: https://github.com/carlocab/workflow-test/runs/7612001250?check_suite_focus=true#step:7:30

**Comment by fxcoudert** [comment]
I suggest filing it as a bug with openblas, with this info (working and failing binutils versions) they should be able to repro and isolate the issue. Might be a binutils bug, might also be a weird use case that got broken/mishandled somehow.

**Comment by carlocab** [comment]
https://github.com/xianyi/OpenBLAS/issues/3708

**Comment by carlocab** [comment]
I see this even from building binutils from source outside of Homebrew: https://github.com/carlocab/workflow-test/actions/runs/2776173970

**Comment by carlocab** [comment]
Disabling `DYNAMIC_ARCH` worked. I'm going to do that.

**Pull request closed without merging**